sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+      // Sets key class to key translation function's output class type.
+      return toBuilder().setKeyTranslationFunction(function)
+          .setKeyClass((TypeDescriptor<K>) function.getOutputTypeDescriptor()).build();
+    }
+
+    /**
+     * Returns a new {@link HadoopInputFormatIO.Read} that will transform the values read from the
+     * source using the given value translation function.
+     *
+     * <p>
+     * Does not modify this object.
+     */
+    public Read<K, V> withValueTranslation(SimpleFunction<?, V> function) {
+      checkNotNull(function, HadoopInputFormatIOContants.NULL_VALUE_TRANSLATIONFUNC_ERROR_MSG);
+      // Sets value class to value translation function's output class type.
+      return toBuilder().setValueTranslationFunction(function)
+          .setValueClass((TypeDescriptor<V>) function.getOutputTypeDescriptor()).build();
+    }
+
+    @Override
+    public PCollection<KV<K, V>> expand(PBegin input) {
+      HadoopInputFormatBoundedSource<K, V> source = new HadoopInputFormatBoundedSource<K, V>(
+          this.getConfiguration(),
+          this.getKeyCoder(),
+          this.getValueCoder(),
+          this.getKeyTranslationFunction(),
+          this.getValueTranslationFunction(),
+          null);
+      return input.getPipeline().apply(org.apache.beam.sdk.io.Read.from(source));
+    }
+
+    /**
+     * Validates inputs provided by the pipeline user before reading the data.
+     */
+    @Override
+    public void validate(PBegin input) {
+      checkNotNull(this.getConfiguration(),
+          HadoopInputFormatIOContants.MISSING_CONFIGURATION_ERROR_MSG);
+      // Validate that the key translation input type must be same as key class of InputFormat.
+      validateTranslationFunction(inputFormatKeyClass, getKeyTranslationFunction(),
+          HadoopInputFormatIOContants.WRONG_KEY_TRANSLATIONFUNC_ERROR_MSG);
+      // Validate that the value translation input type must be same as value class of InputFormat.
+      validateTranslationFunction(inputFormatValueClass, getValueTranslationFunction(),
+          HadoopInputFormatIOContants.WRONG_VALUE_TRANSLATIONFUNC_ERROR_MSG);
+      getKeyAndValueCoder(input);
+    }
+
+
+    /**
+     * Validates translation function given for key/value translation.
+     */
+    private void validateTranslationFunction(TypeDescriptor<?> inputType,
+        SimpleFunction<?, ?> simpleFunction, String errorMsg) {
+      if (simpleFunction != null) {
+        if (!simpleFunction.getInputTypeDescriptor().equals(inputType)) {
+          throw new IllegalArgumentException(
+              String.format(errorMsg, inputFormatClass.getRawType(), inputType.getRawType()));
+        }
+      }
+    }
+
+    private Coder<K> keyCoder;
+    private Coder<V> valueCoder;
+
+    /**
+     *  Sets the key and value coder based on the key class and value class.
+     */
+    protected void getKeyAndValueCoder(PBegin input) {
```

Comment:
we do not need to call this in validate() hence moved it to expand as explained in above comment.